Mathlib/Data/Bitvec/Lemmas.lean

```diff
@@ -1,173 +1,115 @@
+
+@[simp]
+lemma toNat_ofFin (x : Fin (2^w)) : (ofFin x).toNat = x.val := rfl
+
+theorem toNat_append {msbs : BitVec w} {lsbs : BitVec v} :
+    (msbs ++ lsbs).toNat = msbs.toNat <<< v ||| lsbs.toNat := by
+  rcases msbs with ⟨msbs, hm⟩
+  rcases lsbs with ⟨lsbs, hl⟩
+  simp only [HAppend.hAppend, append, toNat_ofFin]
+  rw [toNat_ofNat (Nat.add_comm w v ▸ append_lt hl hm)]
+
+#noalign bitvec.bits_to_nat_to_bool
+#align bitvec.to_nat_of_nat Std.BitVec.toNat_ofNat
+
+@[simp]
+lemma extractLsb_eq {w : ℕ} (hi lo : ℕ) (a : BitVec w) :
+    extractLsb hi lo a = extractLsb' lo (hi - lo + 1) a :=
   rfl
-#align bitvec.of_nat_succ Bitvec.ofNat_succ
 
-theorem toNat_ofNat {k n : ℕ} : Bitvec.toNat (Bitvec.ofNat k n) = n % 2 ^ k := by
-  induction' k with k ih generalizing n
-  · simp [Nat.mod_one]
-    rfl
-  · rw [ofNat_succ, toNat_append, ih, bits_toNat_decide, mod_pow_succ, Nat.mul_comm]
-#align bitvec.to_nat_of_nat Bitvec.toNat_ofNat
+theorem toNat_extractLsb' {i j} {x : BitVec w} :
+    (extractLsb' i j x).toNat = x.toNat / 2 ^ i % (2 ^ j) := by
+  simp only [extractLsb', ofNat_eq_mod_two_pow, shiftRight_eq_div_pow]
+
+theorem getLsb_eq_testBit {i} {x : BitVec w} : getLsb x i = x.toNat.testBit i := by
+  simp only [getLsb, Nat.shiftLeft_eq, one_mul, Nat.and_two_pow]
+  cases' testBit (BitVec.toNat x) i
+  <;> simp [pos_iff_ne_zero.mp (two_pow_pos i)]
 
 theorem ofFin_val {n : ℕ} (i : Fin <| 2 ^ n) : (ofFin i).toNat = i.val := by
-  rw [ofFin, toNat_ofNat, Nat.mod_eq_of_lt]
-  apply i.is_lt
-#align bitvec.of_fin_val Bitvec.ofFin_val
+  rfl
+#align bitvec.of_fin_val Std.BitVec.ofFin_val
 
 theorem addLsb_eq_twice_add_one {x b} : addLsb x b = 2 * x + cond b 1 0 := by
-  simp [addLsb, two_mul]
-#align bitvec.add_lsb_eq_twice_add_one Bitvec.addLsb_eq_twice_add_one
-
-theorem toNat_eq_foldr_reverse {n : ℕ} (v : Bitvec n) :
-    v.toNat = v.toList.reverse.foldr (flip addLsb) 0 := by rw [List.foldr_reverse]; rfl
-#align bitvec.to_nat_eq_foldr_reverse Bitvec.toNat_eq_foldr_reverse
-
-theorem toNat_lt {n : ℕ} (v : Bitvec n) : v.toNat < 2 ^ n := by
-  suffices v.toNat + 1 ≤ 2 ^ n by simpa
-  rw [toNat_eq_foldr_reverse]
-  cases' v with xs h
-  dsimp [Bitvec.toNat, bitsToNat]
-  rw [← List.length_reverse] at h
-  generalize xs.reverse = ys at h
-  induction' ys with head tail ih generalizing n
-  · simp [← h]
-  · simp only [← h, pow_add, flip, List.length, List.foldr, pow_one]
-    rw [addLsb_eq_twice_add_one]
-    trans 2 * List.foldr (fun (x : Bool) (y : ℕ) => addLsb y x) 0 tail + 2 * 1
-    -- Porting note: removed `ac_mono`, `mono` calls
-    · rw [add_assoc]
-      apply Nat.add_le_add_left
-      cases head <;> decide
-    · rw [← left_distrib]
-      rw [mul_comm _ 2]
-      apply Nat.mul_le_mul_left
-      exact ih rfl
-#align bitvec.to_nat_lt Bitvec.toNat_lt
+  simp [addLsb, two_mul]; cases b <;> rfl
+#align bitvec.add_lsb_eq_twice_add_one Std.BitVec.addLsb_eq_twice_add_one
+
+#noalign bitvec.to_nat_eq_foldr_reverse
```

Comment:
Let's leave a breadcrumb
```suggestion
-- The previous statement was `(v : Bitvec n) : v.toNat = v.toList.reverse.foldr (flip addLsb) 0`.
-- Since the statement is awkward and `Std.BitVec` has no comparable API, we just drop it.
#noalign bitvec.to_nat_eq_foldr_reverse
```